### ISSUE TITLE
feature_macro test to stringent for spirv and binary modes

### ISSUE THREAD

**Issue by FranklandJack** [KhronosGroup/OpenCL-CTS#1026]
The feature macro test: https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/compiler/test_feature_macro.cpp calls `create_single_kernel_helper_create_program` which in binary and spirv mode will call into offline compiler executables. 

In the case of a compilation error `create_single_kernel_helper_create_program` returns an error to the caller in binary/spirv mode since the compiler will not have produced and output file it can read back into the runtime (this doesn't happen in online mode since the CL C isn't compiled during the call to `create_single_kernel_helper_create_program` in this case, this happens later on in a call to `clBuildProgram`). 

Since this feature macro test expects the compiler to fail (it uses this information later on to detect whether a feature macro is supported by the compiler) the return error code of these calls to `create_single_kernel_helper_create_program` should not cause the test itself to fail in spirv/binary modes like it does here: https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/compiler/test_feature_macro.cpp#L143.

As it stands the test is skipped in binary mode, but is not in spirv mode since there needs to be a compiler in the runtime to consume the spirv. By replacing calls to `create_single_kernel_helper_create_program` with calls to `create_single_kernel_helper` and removing subsequent calls to `clBuildProgram` (like here: https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/compiler/test_feature_macro.cpp#L143) in the test, then using the results of the calls to `create_single_kernel_helper` to determine whether compilation succeeded or failed (this will work for all three compilation modes), the above issue could be resolved and this test could be run in binary mode as well as online and spirv.

**Comment by StuartDBrady** [comment]
I'm not especially familiar with the feature macro test in question, so my apologies if my next question is somewhat naïve.

By doing this, would this mean that we'd simply be testing the offline compiler's support of feature macros, and not testing the online compiler? If so, is there value directly related to conformance in testing of the offline compiler in isolation? I thought the point of offline compilation was to test the _consumption_ of binaries and of SPIR-V, not the generation of them. It certainly seems like useful testing to do, but I am not sure this would be testing conformance per se.

The point that for online compilation, `create_single_kernel_helper_create_program()` doesn't result in any compilation being performed seems notable. Has anyone looked into refactoring the CTS to reduce its exposure to this difference between online and offline compilation modes? It may also be worth noting that another area with room for improvement is the handling of multiple devices with offline compilation.

**Comment by FranklandJack** [comment]
> I'm not especially familiar with the feature macro test in question, so my apologies if my next question is somewhat naïve.
> 
> By doing this, would this mean that we'd simply be testing the offline compiler's support of feature macros, and not testing the online compiler? If so, is there value directly related to conformance in testing of the offline compiler in isolation? I thought the point of offline compilation was to test the _consumption_ of binaries and of SPIR-V, not the generation of them. It certainly seems like useful testing to do, but I am not sure this would be testing conformance per se.
> 
> The point that for online compilation, `create_single_kernel_helper_create_program()` doesn't result in any compilation being performed seems notable. Has anyone looked into refactoring the CTS to reduce its exposure to this difference between online and offline compilation modes? It may also be worth noting that another area with room for improvement is the handling of multiple devices with offline compilation.

Thanks for taking the time to review this issue @StuartDBrady, it is really appreciated.

I think in the original issue description I conflated the notions of `binary` mode and running with a compiler-less driver, so that might have caused some confusion, apologies.

Your question is a really good one, to answer the first part:
> By doing this, would this mean that we'd simply be testing the offline compiler's support of feature macros, and not testing the online compiler?

I believe that since the user specifies the mode to run the CTS, either `online`, `binary` or `spir-v` we would still be testing the online compiler in `online` mode since in that case the call to `create_single_kernel_helper` in the suggested change would resolve to the runtime compiler. Currently this test **is** testing that the feature macros as defined by the offline compiler when run in `binary` and `spir-v` mode are correct and match the features reported in the runtime, it's just that the test uses the failure of the compiler to determine whether the features macros were defined or not and in the case of `binary` and `spir-v` modes, this causes the whole test to fail. 

And to answer the second part of the question:
>If so, is there value directly related to conformance in testing of the offline compiler in isolation? I thought the point of offline compilation was to test the consumption of binaries and of SPIR-V, not the generation of them. It certainly seems like useful testing to do, but I am not sure this would be testing conformance per se.


This test is only skipped when no compiler is detected in the runtime, if run in `binary` or `spir-v` mode it will still shell out to the offline compiler and not use the runtime compiler. If we think this test should only be run with an online compiler, then perhaps it should only be run in `online` mode, not whenever a compiler is present in the driver. I think since this test is checking that the feature macros defined in the online/offline compiler match the features reported in the driver (not just the feature macros in the online/offline compiler), it's a useful test to run in `binary` and `spir-v` modes.

>The point that for online compilation, create_single_kernel_helper_create_program() doesn't result in any compilation being performed seems notable. Has anyone looked into refactoring the CTS to reduce its exposure to this difference between online and offline compilation modes? It may also be worth noting that another area with room for improvement is the handling of multiple devices with offline compilation.

I definitely agree, I think if this test is going to be called in any of the `binary` or `spir-v` modes as well as `online`, then the this is required. The API I suggested: `create_single_kernel_helper` would do this I think.

Let me know if this answers your questions or if you have anymore.

**Comment by StuartDBrady** [comment]
> I believe that since the user specifies the mode to run the CTS, either `online`, `binary` or `spir-v` we would still be testing the online compiler in `online` mode since in that case the call to `create_single_kernel_helper` in the suggested change would resolve to the runtime compiler. Currently this test **is** testing that the feature macros as defined by the offline compiler when run in `binary` and `spir-v` mode are correct and match the features reported in the runtime, it's just that the test uses the failure of the compiler to determine whether the features macros were defined or not and in the case of `binary` and `spir-v` modes, this causes the whole test to fail.

I see. So it's useful to confirm that the compiler matches the runtime, because other CTS tests will rely upon this without necessarily catching any mismatch?

In that case I would agree that this is needed.

Regarding `create_single_kernel_helper()`, that is generally what I would recommend, although there are cases in other tests where it is not appropriate, and again, multi-device support is a generally a concern.

**Comment by StuartDBrady** [comment]
(Apologies, misclicked.)

**Comment by FranklandJack** [comment]
> I see. So it's useful to confirm that the compiler matches the runtime, because other CTS tests will rely upon this without necessarily catching any mismatch?

Yes. I think the tests in `test_conformance/compiler/test_feature_macro.cpp` are checking that the feature macro is defined in the compiler *if and only if* the corresponding runtime query returns that the feature is supported. 

For example, to test support for pipes, this test first queries the runtime with `clGetDeviceInfo` called with the `CL_DEVICE_PIPE_SUPPORT` query, then calls to a compiler (offline or online) with two kernels using the `__opencl_c_pipes` feature macro to determine whether the macro is supported or not. The test then only passes if both query and compiler report that it isn't supported or both report that it is.

**Comment by StuartDBrady** [comment]
That raises the question of how this would interact with a prepopulated offline compilation cache, in cases where there are cache misses (likely due to changes in CL source, or due to use of a partial cache). What if the offline compiler in use is inconsistent with the runtime, but this doesn't get detected because the content in the prepopulated cache for the feature macros test _is_ consistent with the runtime? Conversely, what if the prepopulated cache is inconsistent with the runtime, but this doesn't get detected because there are cache misses in the feature macros test?

Should we (or do we?) discourage use of prepopulated offline compilation caches in certain cases?

**Comment by FranklandJack** [comment]
This is a really good point, but I think it might be beyond the scope of this issue. As far as I understand it any changes to an ICD + compiler pair being tested, or to the CTS test suite itself would constitute a cache invalidation, and it would be the users responsibility to clear their cache to force consistency in this case.

**Comment by StuartDBrady** [comment]
Provided that we agree that it is the user's responsibility to clear the cache, I agree that my question is beyond the scope of this issue. If we don't agree that it is the user's responsibility, then that calls into question the benefit of resolving this issue.

Would it increase the scope hugely to ask that the test harness check the compilation cache at startup and emit a warning if it is not empty? I think that would allow this issue about consistency between the compiler and the runtime to be resolved with greater confidence.

**Comment by FranklandJack** [comment]
Good point. I think maybe the most pragmatic solution to this is to being by simply adding this test to the list of those that should get skipped in `binary` and `spir-v` mode. Then we can discuss re-adding it in offline modes once the above points have been addressed.
